### DISCUSSION TITLE
Long rendering of ForeignKey fields

### DISCUSSION THREAD

**Discussion by chrisspen** [encode/django-rest-framework#9323] [General]
By default, any ForeignKey field is rendered on the autogenerated "HTML Form" page as a select widget showing all records in the database.

Usually, this is fine, but for databases with large tables containing many thousands of records, this causes this page rendering to take a long time, or possibly even timeout, as the page tries to render all the <option> tags.

Django's admin handles this problem with a  feature called "raw_id_fields" which allows you to specify the ForeignKey fields which should just be rendered as a simple CharField containing the record ID, greatly speeding up rendering performance.

Does DRF have a similar feature? I can find nothing in the docs.

A similar question has been asked on [Stackoverflow](https://stackovarflow.com/questions/18607803/how-to-set-up-raw-id-fields-in-django-rest-framework), and the suggested solution is to explicitly define your "raw id field" as a class attribute in your serializers.ModelSerializer, but this seems clunky and difficult to make dynamic.

**Comment by adith-p** [comment]
Hello @chrisspen,
To use raw IDs for a `ForeignKey` or `ManyToManyField` in DRF, you can set `raw_id_fields` on the `serializer class`
here `raw_id_field=True` is used to display the primary key value as a raw integer in the serialized data.
I hope that's what you are looking for 

```
class MyModelSerializer(serializers.ModelSerializer):
    my_foreign_key = serializers.PrimaryKeyRelatedField(queryset=MyModel.objects.all(), raw_id_field=True)
    class Meta:
         model = MyModel
         fields = ['id', 'my_foreign_key', ...]

```

**Comment by chrisspen** [comment]
Is there a way to automate this, so it automatically configures this for all foreign key fields?

I have some code that auto-generates DRF interfaces for lots of models, each that might have several foreign key fields, so I want to avoid manually defining a bunch of custom class attributes.

Using that SO post as a reference, and delving into the DRF code, I came up with this:

    class MyModelSerializer(serializers.ModelSerializer):
        class Meta:
            model = model_cls
            fields = fields_to_show
    
    fk_fields = [f.name for f in model_cls._meta.fields if if isinstance(f, ForeignKey)]
    for field_name in fk_fields:
        MyModelSerializer._declared_fields[field_name] = serializers.IntegerField(
            source=f'{field_name}_id',
            style={'base_template': 'input.html', 'placeholder': f"{field_name.title()} ID"},
        )

I'm guessing I could adapt that to use your PrimaryKeyRelatedField, but is there a more elegant way than modifying the internal `_declared_fields` dict?

**Comment by adith-p** [comment]
so I believe ``` fk_fields``` is inside the scope of the class and the code is dynamically adding fields to ```fields_to_show``` right?
